### PR TITLE
Add support for projectiles and stun effect (wip)

### PR DESCRIPTION
**Pull request by matiascalegaris**

### CODE REVIEW

File: Codigo/Declares.bas [hunk 1]
```diff
@@ -969,7 +971,7 @@ Public Type t_UserOBJ
     ObjIndex As Integer
     amount As Integer
     Equipped As Byte
-
+    LastUseTime As Long
```
Comment (morgolock) [inline Codigo/Declares.bas:974]: Rename to something better, `LastUseTime` does not mention what is to be used...

File: Codigo/InvUsuario.bas [hunk 2]
```diff
@@ -3434,14 +3440,16 @@ Function ItemNewbie(ByVal ItemIndex As Integer) As Boolean
...
 
-        
         Exit Function
 
 ItemNewbie_Err:
 102     Call TraceError(Err.Number, Err.Description, "InvUsuario.ItemNewbie", Erl)
+End Function
 
-        
+Public Function IsItemInCd(ByRef obj As t_UserOBJ) As Boolean
```
Comment (morgolock) [inline Codigo/InvUsuario.bas:3451]: Please rename to `IsItemInCooldown`

File: Codigo/MODULO_NPCs.bas [hunk 3]
```diff
@@ -1824,4 +1824,14 @@ Public Sub KillRandomNpc()
...
+Public Function CanMove(counter As t_NpcCounters, flags As t_NPCFlags) As Boolean
+    CanMove = flags.Inmovilizado + flags.Paralizado = 0 And counter.StunEndTime < GetTickCount()
+End Function
+
+Public Function CanAttack(counter As t_NpcCounters, flags As t_NPCFlags) As Boolean
+    CanAttack = flags.Inmovilizado + flags.Paralizado = 0 And counter.StunEndTime < GetTickCount()
+End Function
 
+Public Function StunNPc(ByRef Counters As t_NpcCounters)
+    Counters.StunEndTime = GetTickCount() + 5000
```
Comment (morgolock) [inline Codigo/MODULO_NPCs.bas:1836]: replace 5000 by `NPC_STUN_TIME`

File: Codigo/Protocol_Writes.bas [hunk 4]
```diff
@@ -3334,6 +3334,16 @@ WriteParalizeOK_Err:
         '</EhFooter>
 End Sub
 
+Public Sub WriteStunStart(ByVal UserIndex As Integer, duration As Integer)
+    On Error GoTo WriteStunStart_Err
+100     Call Writer.WriteInt16(ServerPacketID.StunStart)
+        Call Writer.WriteInt16(duration)
+102     Call modSendData.SendData(ToIndex, UserIndex)
```
Comment (morgolock) [inline Codigo/Protocol_Writes.bas:3341]: missing `exit sub`

File: Codigo/SistemaCombate.bas [hunk 5]
```diff
@@ -1002,15 +1032,15 @@ NpcAtacaNpc_Err:
         
 End Sub
 
-Public Sub UsuarioAtacaNpc(ByVal UserIndex As Integer, ByVal NpcIndex As Integer)
+Public Sub UsuarioAtacaNpc(ByVal UserIndex As Integer, ByVal npcIndex As Integer, aType As AttackType)
```
Comment (morgolock) [inline Codigo/SistemaCombate.bas:1035]: `byval aType As AttackType`

File: Codigo/SistemaCombate.bas [hunk 6]
```diff
@@ -573,7 +579,7 @@ CalcularDaño_Err:
         
 End Function
 
-Private Sub UserDañoNpc(ByVal UserIndex As Integer, ByVal NpcIndex As Integer)
+Private Sub UserDañoNpc(ByVal UserIndex As Integer, ByVal npcIndex As Integer, aType As AttackType)
```
Comment (morgolock) [inline Codigo/SistemaCombate.bas:582]: `byval aType As AttackType`

File: Codigo/SistemaCombate.bas [hunk 7]
```diff
@@ -363,7 +368,7 @@ PoderAtaqueWrestling_Err:
         
 End Function
 
-Private Function UserImpactoNpc(ByVal UserIndex As Integer, ByVal NpcIndex As Integer) As Boolean
+Private Function UserImpactoNpc(ByVal UserIndex As Integer, ByVal npcIndex As Integer, aType As AttackType) As Boolean
```
Comment (morgolock) [inline Codigo/SistemaCombate.bas:371]: `byval aType As AttackType`

File: Codigo/SistemaCombate.bas [hunk 8]
```diff
@@ -1193,7 +1223,7 @@ UsuarioAtaca_Err:
         
 End Sub
 
-Private Function UsuarioImpacto(ByVal AtacanteIndex As Integer, ByVal VictimaIndex As Integer) As Boolean
+Private Function UsuarioImpacto(ByVal AtacanteIndex As Integer, ByVal VictimaIndex As Integer, aType As AttackType) As Boolean
```
Comment (morgolock) [inline Codigo/SistemaCombate.bas:1226]: `byval aType As AttackType`

File: Codigo/SistemaCombate.bas [hunk 9]
```diff
@@ -1290,7 +1320,7 @@ UsuarioImpacto_Err:
 
 End Function
 
-Public Sub UsuarioAtacaUsuario(ByVal AtacanteIndex As Integer, ByVal VictimaIndex As Integer)
+Public Sub UsuarioAtacaUsuario(ByVal AtacanteIndex As Integer, ByVal VictimaIndex As Integer, aType As AttackType)
```
Comment (morgolock) [inline Codigo/SistemaCombate.bas:1323]: `byval aType As AttackType`

File: Codigo/SistemaCombate.bas [hunk 10]
```diff
@@ -1341,7 +1371,7 @@ UsuarioAtacaUsuario_Err:
 
 End Sub
 
-Private Sub UserDañoUser(ByVal AtacanteIndex As Integer, ByVal VictimaIndex As Integer)
+Private Sub UserDañoUser(ByVal AtacanteIndex As Integer, ByVal VictimaIndex As Integer, aType As AttackType)
```
Comment (morgolock) [inline Codigo/SistemaCombate.bas:1374]: `byval aType As AttackType`

File: Codigo/SistemaCombate.bas [hunk 11]
```diff
@@ -2348,7 +2378,7 @@ ErrHandler:
 
 End Function
 
-Private Sub UserDañoEspecial(ByVal AtacanteIndex As Integer, ByVal VictimaIndex As Integer)
+Private Sub UserDañoEspecial(ByVal AtacanteIndex As Integer, ByVal VictimaIndex As Integer, aType As AttackType)
```
Comment (morgolock) [inline Codigo/SistemaCombate.bas:2381]: `byval aType As AttackType`

**Review by morgolock — CHANGES_REQUESTED** [review]

File: Codigo/Declares.bas (hunk 1 above)
Comment (matiascalegaris) [reply Codigo/Declares.bas:974]: its the last use time of the object, its inside a t_UserObj so it refers to the last time this object was used

**Review by morgolock — APPROVED** [review]

**Pull request merged**